streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -454,6 +479,15 @@ private void handleTasksWithoutStateUpdater(final Map<TaskId, Set<TopicPartition
                                                 final Map<TaskId, Set<TopicPartition>> standbyTasksToCreate,
                                                 final Map<Task, Set<TopicPartition>> tasksToRecycle,
                                                 final Set<Task> tasksToCloseClean) {
+        final Map<Task, Set<TopicPartition>> pendingStandbyTasksToRecycle = assignPendingTasks(activeTasksToCreate, logPrefix, topologyMetadata, changelogReader);
+        final Map<Task, Set<TopicPartition>> pendingStandbyTasksToUse = assignPendingTasks(standbyTasksToCreate, logPrefix, topologyMetadata, changelogReader);
+
+        // recycle the pending standbys to active
+        tasks.addStandbyTasks(pendingStandbyTasksToRecycle.keySet());
+
+        // use pending Standbys as real Standby tasks
+        tasks.addStandbyTasks(pendingStandbyTasksToUse.keySet());
```

Comment:
If you have multiple StreamThreads on the same instance, they share the set of "pending tasks". `assignPendingTasks` actually _removes_ the tasks from the shared pending tasks, preventing other threads from using them.

The only way to do as you suggest, would not removing the tasks from the shared set of pending tasks until the end of the assignment process, at which point we would need to add a new step to go through the assigned tasks and remove any from the shared pending task set. I think this is possible, but I'm not convinced it would be any simpler than the current solution. I can experiment with this to see what it's like though.

Reply:
Baseline: I missed something!
You are right! With multiple stream threads my proposal is not that simple anymore.

Reply:
@cadonna Do you mind if we leave it as-is then? Or do you think there's a better way to do it?

Reply:
Let's leave it as you did.